vetomint/src/progress.rs

```diff
@@ -84,54 +328,160 @@ pub(super) fn progress(
+                        responses.append(&mut on_timeout_propose(state));
+                    }
+                }
             }
+            ConsensusStep::Precommit => {
+                if let Some(timeout_precommit) = state.timeout_precommit {
+                    if time >= timeout_precommit {
+                        responses.append(&mut on_timeout_precommit(state, time));
+                    }
+                }
+            }
+            _ => (),
         }
         Some(responses)
     }
 }
 
-enum StartRoundResponse {
-    Normal(Vec<ConsensusResponse>),
-    /// Emits a `CreateProposal`.
-    Pending,
-}
-
 fn start_round(
-    height_info: &HeightInfo,
     state: &mut ConsensusState,
     round: usize,
     time: Timestamp,
-) -> StartRoundResponse {
+) -> Vec<ConsensusResponse> {
     state.round = round;
     state.step = ConsensusStep::Propose;
-    let proposer = decide_proposer(round, height_info);
-    if proposer == height_info.this_node_index {
-        if let Some(valid_value) = state.valid_value {
-            StartRoundResponse::Normal(vec![ConsensusResponse::BroadcastProposal {
-                proposal: valid_value,
-                round,
-            }])
+    state.timeout_precommit = None;
+    let proposer = decide_proposer(round, &state.height_info);
+    if Some(proposer) == state.height_info.this_node_index {
+        let proposal = if state.valid_value != None {
+            state.valid_value.unwrap()
         } else {
-            StartRoundResponse::Pending
+            state.waiting_for_proposal_creation = true;
+            state.block_candidate
+        };
+        vec![ConsensusResponse::BroadcastProposal { proposal, round }]
+    } else {
+        state.timeout_propose = Some(time + state.height_info.consensus_params.timeout_ms as i64);
+        Vec::new()
+    }
+}
+
+fn on_proposal(
+    proposal: BlockIdentifier,
+    favor: bool,
+    state: &mut ConsensusState,
+    round: Round,
+) -> Vec<ConsensusResponse> {
+    let this_node_voting_power = if state.height_info.this_node_index == None {
+        0
+    } else {
+        state.height_info.validators[state.height_info.this_node_index.unwrap()]
+    };
+    let mut response: Vec<ConsensusResponse> = Vec::new();
+
+    state.step = ConsensusStep::Prevote;
+    state.votes.insert(round, {
+        let mut votes = state
+            .votes
+            .get(&round)
+            .unwrap_or(&Default::default())
+            .clone();
+        votes.prevotes_total += this_node_voting_power;
+        votes.precommits_total += this_node_voting_power;
+        if favor {
+            votes.prevotes_favor.insert(
+                proposal,
+                votes.prevotes_favor.get(&proposal).unwrap_or(&0) + this_node_voting_power,
+            );
+            votes.precommits_favor.insert(
+                proposal,
+                votes.precommits_favor.get(&proposal).unwrap_or(&0) + this_node_voting_power,
+            );
```

Comment:
`prevotes_total` means total voting power voted in that round. And `prevotes_favor` has BTreeMap structure so that can stores voting power voted to multiple proposals in same round.